RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,566 @@
+
+  * `version` element (Integer) (**Required**)
+
+  Which version of the templating engine this module uses. If the template specifies a version that the templating engine does not support it is expected the engine will raise a terminating error.
+
+  Note that the type here is an integer. Although we could use a semver (Semantic Versioning) based string here, that would make parsing difficult.
+
+* `template_sources` element (Array of String)
+
+  A list of places to find Template Repositories. Order is important. If the same template (specifically the same template directory name) is specified in two sources, then the top-most source will be used ("First one wins")
+
+  Support source types:
+  * Filesystem based repository
+
+    A directory on disk can be specified as a Template Repository.
+
+  * Git based repository
+
+    A git repository (For example Github, GitLab or in internal git repo)
+
+  * `default`
+
+    This special source type will use whatever the Template Engine considers the default.  In the case of the PDK, this would be the cached `pdk-templates` repo that is packaged as part of the PDK. Other tools using the Template Engine may specify a different location.
+
+  If no template sources are specified `default` is used.
+
+* `templates` element (Array of string). List of templates (The template directory name) to be applied **IN ORDER** (Order is important.  See [Resolution Order](#resolution_order))
+
+Example V1 `.sync.yml`
+
+``` yaml
+---
+Gemfile:
+  # Gemfile template customisations
+
+.rubocop.yml:
+  # Rubocop template customisations
+
+# ... etc
+```
+
+Example `.sync.yml` with V2 settings
+
+``` yaml
+---
+pdk_template:
+  version: 2
+  template_sources:
+    -
+      type: git
+      location: 'https://github.com/user/custom-templates.git'
+    - default
+  templates:
+    - appveyor_ci
+    - travis_ci
+    - rubocop
+    - litmus_tests
+
+gemfile:
+  # Gemfile template customisations
+
+rubocop:
+  # Rubocop template customisations
+
+# ... etc
+```
+
+#### Filesystem based template sources
+
+Example:
+
+``` yaml
+---
+pdk_template:
+  template_sources:
+    -
+      type: filesystem
+      location: 'C:/templates/custom'
+    #  Unix style path
+    # -
+    #   type: filesystem
+    #   location: '/templates/custom'
+```
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| type         | Must be `filesystem` |
+| location     | The path the repository directory |
+
+#### Git based template sources
+
+``` yaml
+---
+pdk_template:
+  template_sources:
+    -
+      type: git
+      location: 'https://github.com/user/custom-templates.git'
+      ref: master
+```
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| type         | Must be `git` |
+| location     | The git remote to clone |
+| ref          | (Optional) The git reference to checkout. Default is what the git remote deems the default branch |
```

Comment:
Do we need a special value that can be used to mimic the default template behavior of looking for a tag matching the PDK version number?